### DISCUSSION TITLE
Confusing execution order, `always` is seemingly before everything

### DISCUSSION THREAD

**Discussion by n8jadams** [statelyai/xstate#1438] [General] [answered]
Consider the following machine:

```javascript
import { Machine, interpret } from 'xstate'

const machine = Machine({
  id: 'test-machine',
  initial: 'stepOne',
  context: { error: false },
  states: {
    stepOne: {
      exit: [() => {console.log('stepOne - exit')}]
    },
    beforeStepTwo: {
      entry: [
        () => {console.log('beforeStepTwo - entry')},
        'validate-step-one'
      ],
      always: [
        {
          target: 'error',
          cond: ({ error }) => {
            console.log('beforeStepTwo - always')
            return error
          }
        },
        { target: 'stepTwo' }
      ]
    },
    stepTwo: {},
    error: {}
  },
  on: {
    CONTINUE_TO_STEP_TWO: 'beforeStepTwo'
  }
}, {
  actions: {
    'validate-step-one': () => {
      return {
        error: true
      }
    }
  }
})

const interpretedMachine = interpret(machine).start()
interpretedMachine.send('CONTINUE_TO_STEP_TWO')
```

This results in the following being logged:
```
beforeStepTwo - always
stepOne - exit
beforeStepTwo - entry 
```

I find this very strange. Why is `always` happening first, before the `exit` of the previous state, and the `entry` of the same state? My goal is to have a transition happen, and have the `'validate-step-one'` occur BEFORE the condition check in the `always` transitions.

Is my only option to add another state? So should I do something like this:

```
stepOne -> beforeBeforeStepTwo (which runs the 'validate-step-one' action) -> beforeStepTwo (which has the always conditions)

**Comment by Andarist** [comment, marked as answer]
Currently, custom actions (both of your `beforeStepTwo.entry` actions are custom) are executed **after** the machine settles a particular macro transition. OTOH, guards are always executed within the macrostep as they are needed to determine what transitions should be taken.

I totally understand why this might be confusing. This issue seems to have popped up several times recently.

Your example seems to be a little bit broken though - shouldn't `validate-step-one` be an `assign` action?

**Reply by n8jadams** [reply]
Aha! Yes, `validate-step-one` should've been an `assign` action. Then everything works as expected and in the order I expect.

Thanks!
